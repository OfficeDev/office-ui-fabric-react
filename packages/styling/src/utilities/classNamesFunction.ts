import { mergeStyleSets } from '../MergeStyles';
import {
  IClassNames,
  IStyleFunction
} from '../interfaces/index';

/**
 * Creates a getClassNames function which calls getStyles given the props, and injects them
 * into mergeStyleSets.
 *
 * @public
 */
export function classNamesFunction<TStyleProps extends {}, TStyles extends {}>(): (
  getStyles?: IStyleFunction<TStyleProps, TStyles>,
  styleProps?: TStyleProps
) => IClassNames<TStyles> {

  return (
    getStyles?: IStyleFunction<TStyleProps, TStyles>,
    styleProps?: TStyleProps
  ): IClassNames<TStyles> => mergeStyleSets(getStyles && getStyles(styleProps!)
  );
}
