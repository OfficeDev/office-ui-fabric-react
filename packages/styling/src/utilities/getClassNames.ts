import { css } from 'glamor';

export type IClassNames<T> = {
  [P in keyof T]?: string;
};

export function getClassNames<T>(styles: T): IClassNames<T> {
  return Object
    .keys(styles)
    .reduce((classNames: IClassNames<T>, className: string) => {
      // tslint:disable-next-line:no-any
      classNames[className] = css((<any>styles)[className]).toString();
      return classNames;
    }, {});
}
