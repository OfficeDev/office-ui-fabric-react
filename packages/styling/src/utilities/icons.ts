import {
  GlobalSettings
} from '@uifabric/utilities/lib/GlobalSettings';
import {
  IExtendedRawStyle,
  IFontFace,
  fontFace,
  mergeStyles
} from '@uifabric/merge-styles';

export interface IIconSubset {
  fontFace: IFontFace;
  icons: {
    [key: string]: string;
  };
  style?: IExtendedRawStyle;
}

export interface IIconSubsetRecord extends IIconSubset {
  isRegistered?: boolean;
  className?: string;
}

export interface IIconRecord {
  code: string;
  subset: IIconSubsetRecord;
  className?: string;
}

export interface IIconRecords {
  [key: string]: IIconRecord;
}

const ICON_SETTING_NAME = 'icons';
const _icons: IIconRecords = GlobalSettings.getValue(ICON_SETTING_NAME, {});

/**
 * Registers a given subset of icons.
 *
 * @param iconSubset - the icon subset definition.
 */
export function registerIcons(iconSubset: IIconSubset): void {
  let subset = {
    ...iconSubset,
    isRegistered: false,
    className: undefined
  };
  let { icons } = iconSubset;

  for (const iconName in icons) {
    if (icons.hasOwnProperty(iconName)) {
      const code = icons[iconName];

      _icons[iconName] = {
        code,
        subset,
        className: undefined
      };
    }
  }
}

/**
 * Gets an icon definition. If an icon is requested but the subset has yet to be registered,
 * it will get registered immediately.
 *
 * @public
 * @param name - Name of icon.
 */
export function getIcon(name?: string | null): IIconRecord | undefined {
  let icon: IIconRecord = _icons[name!];

  if (icon) {
    let { subset } = icon;

    if (!subset.isRegistered) {
      // Register font face for given icons.
      fontFace(subset.fontFace);

      // Generate a base class name for the given font.
      subset.className = mergeStyles(
        subset.style,
        {
          fontFamily: subset.fontFace.fontFamily,
          fontWeight: subset.fontFace.fontWeight || 'normal',
          fontStyle: subset.fontFace.fontStyle || 'normal'
        }).toString();

      subset.isRegistered = true;
    }
  }

  return icon;
}
