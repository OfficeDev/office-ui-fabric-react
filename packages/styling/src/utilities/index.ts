export {
  mergeStyles
} from './mergeStyles';

export {
  mergeStyleSets
} from './mergeStyleSets';

export {
  buildClassMap
} from './buildClassMap';

export {
  IIconRecord,
  IIconSubset,
  getIcon,
  registerIcons
} from './icons';
