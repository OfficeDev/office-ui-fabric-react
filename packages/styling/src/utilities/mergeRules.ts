import { css } from './css';

export function mergeRules(...args: {[index: string]: Object}[]): Object {
  let styles: {[index: string]: Object} = {};
  let styleToMerge: Object = args[0];

  for (let prop in styleToMerge) {
    if (styleToMerge.hasOwnProperty(prop)) {
      let allArgs: Object[] = [];
      for (let i: number = 0; i < args.length; i++) {
        let currentArg: {[index: string]: Object}= args[i];

        if (currentArg && currentArg[prop]) {
          allArgs.push(currentArg[prop]);
        }
      }
      styles[prop] = css(...allArgs);
    }
  }
  return styles;
}
