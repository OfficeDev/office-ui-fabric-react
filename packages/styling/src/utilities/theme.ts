export interface IColorPalette {
  backgroundColor: string;
  themePrimaryColor: string;
  themeAccentColor: string;
}

export interface IFont {
  fontFamily: string;
  fontSize: string;
  fontWeight: string;
  fontStyle?: string;
}

export interface IFontPallete {
  tiny: IFont;
  xsmall: IFont;
  small: IFont;
  smallPlus: IFont;
  medium: IFont;
  mediumPlus: IFont;
  large: IFont;
  xLarge: IFont;
  xxLarge: IFont;
  superLarge: IFont;
}
/*

export interface ITheme {

}

export function setTheme(theme: ITheme): void {

}

export function generateColorSteps(fromColor: string, toColor: string, stepsInBetween: number) {


}

export function createTheme(colors: IColorPalette, fonts: IFontPallete): ITheme {
  return {
    themePrimary: colors.themePrimaryColor,
    themeSecondary: colors.themeSecondary || lighter(colors.themePrimaryColor, 10),
    themeTertiary
  };
}

export function lighter(color: string) {

}*/
