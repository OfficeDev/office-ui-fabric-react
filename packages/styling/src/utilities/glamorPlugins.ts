import * as Glamor from 'glamor';
import rtlify from 'rtl-css-js';
import { getRTL } from '@uifabric/utilities';

Glamor['plugins'].add(
  ({ selector, style }) => (
    {
      selector,
      style: getRTL() ? rtlify(style) : style
    }
  ));
