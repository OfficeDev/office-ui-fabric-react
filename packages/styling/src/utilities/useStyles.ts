import * as React from 'react';
import { mergeStyleSets, IStyleSet, IProcessedStyleSet } from '@uifabric/merge-styles';
import { ITheme } from '../interfaces/';
import { useCustomizationSettings } from '@uifabric/utilities';
import { getTheme } from '../styles/theme';

export function useStyles<TStyleSet extends IStyleSet<TStyleSet>>(
  styleFunction: (theme: ITheme) => TStyleSet,
): IProcessedStyleSet<TStyleSet> {
  const theme = useCustomizationSettings(['theme']).theme || getTheme();

  return mergeStyleSets(styleFunction?.(theme)) as IProcessedStyleSet<TStyleSet>;
}
