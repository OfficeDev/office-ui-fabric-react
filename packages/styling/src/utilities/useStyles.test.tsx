import * as React from 'react';
import { useStyles } from './useStyles';
import * as renderer from 'react-test-renderer';
import { Customizer } from '@uifabric/utilities';
import { createTheme } from '../styles/theme';

describe('useStyles', () => {
  const Foo = () => {
    const classes = useStyles(theme => ({
      root: {
        background: theme.palette.themePrimary,
      },
    }));

    return <div className={classes.root} />;
  };

  it('can refer to styles from the default theme', () => {
    const component = renderer.create(<Foo />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('can refer to styles from a custom theme', () => {
    const customTheme = createTheme({
      palette: {
        themePrimary: 'purple',
      },
    });

    const component = renderer.create(
      <Customizer settings={{ theme: customTheme }}>
        <Foo />
      </Customizer>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
