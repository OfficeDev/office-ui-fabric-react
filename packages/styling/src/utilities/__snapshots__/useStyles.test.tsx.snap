// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStyles can refer to styles from a custom theme 1`] = `
<div
  className=

      {
        background: purple;
      }
/>
`;

exports[`useStyles can refer to styles from the default theme 1`] = `
<div
  className=

      {
        background: #0078d4;
      }
/>
`;
